### ISSUE TITLE
Gulp format not behaving as expected

### ISSUE THREAD

**Issue by bryceosterhaus** [metal/gulp-metal#26]
This commit https://github.com/metal/metal.js/commit/b467094b6865f01fc51413cdb54464c699929834 is entirely from the `gulp format` task. There are some strange results from the formatter that we should figure out, specifically indentation issues. See https://github.com/metal/metal.js/commit/b467094b6865f01fc51413cdb54464c699929834#diff-8613848b3fa31f88bb889f2483e95900R937 for reference

**Comment by jbalsas** [comment]
Might be worth to take a look into https://github.com/jlongster/prettier for this?

> In technical terms: prettier parses your JavaScript into an AST and pretty-prints the AST, completely ignoring any of the original formatting. Say hello to completely consistent syntax!

**Comment by bryceosterhaus** [comment]
@jbalsas, it looks promising. However, it is very very new so I would be hesitant to adopting something so quickly. I thought the discussion on this issue was very helpful though, https://github.com/jlongster/prettier/issues/40, especially when it comes to all sorts of Liferay formatting standards.

I do love the idea of having something that will auto-format in the editor though, it would be very useful and one less thing to worry about while coding.

**Comment by jbalsas** [comment]
Hey @bryceosterhaus, reading through the updated [Google JavaScript Style Guide](https://google.github.io/styleguide/jsguide.html) I noticed [this appendix](https://google.github.io/styleguide/jsguide.html#appendices-clang-format) in the [9.3 Style-related tools](https://google.github.io/styleguide/jsguide.html#appendices-style-related-tools) section. Seems like we could use [ClangFormat](https://clang.llvm.org/docs/ClangFormat.html) for this.

It has:
- Brew `clang-format` [formula](http://brewformulas.org/ClangFormat)
- [gulp-format](https://www.npmjs.com/package/gulp-format) and [gulp-clang-format](https://www.npmjs.com/package/gulp-clang-format) gulp plugins. The latter seems to be the one used in the Angular project
- [Visual Studio Extension](https://marketplace.visualstudio.com/items?itemName=xaver.clang-format)
- [Sublime Text 3 Extension](https://packagecontrol.io/packages/Clang%20Format)

**Comment by julien** [comment]
Hi @bryceosterhaus and @jbalsas,

I wanted to suggest [ESLint](http://eslint.org) to the possible options, it has a [`--fix`](http://eslint.org/docs/user-guide/command-line-interface#fix) option which can reformat the code too.

**Comment by eduardolundgren** [comment]
To summarize what we've been discussing our JavaScript projects must follow [Google JavaScript Style Guide](https://google.github.io/styleguide/jsguide.html), and it can be enforced by the usage of the [Google eslint config](https://github.com/google/eslint-config-google).

We can use [prettier](https://github.com/prettier/prettier) to auto-format the code based on Google rules (or at least as close as possible).

```js
  gulp.task('format', () => gulp.src('src/**/*.js')
    .pipe(prettier({
      bracketSpacing: false,
      singleQuote: true,
      trailingComma: 'all',
    }))
    .pipe(gulp.dest('src')));
```

Everything seemed fine after formatting with `prettier`, there are few options on ESLint that need to be disabled.

```js
{
  "extends": ["eslint:recommended", "google"],
  "parserOptions": {
      "ecmaVersion": 2017,
      "sourceType": "module"
  },
  "env": {
    "es6": true
  },
  "rules": {
    "arrow-parens": 0, // disabled because of prettier
    "comma-spacing": 0, // disabled because of prettier
    "no-console": 0,
    "no-process-env": 0
  }
}
```

There is one enforcement that `prettier` does that is incompatible with Google Style Guide, the usage of semi-colon. Seems like `prettier` will add support to it https://github.com/prettier/prettier/issues/736. After that we are fine to go with this configuration.

Let me know if you guys are fine with this setup. Thank you!

/cc @fernandosouza @jbalsas Could you guys keep an eye on `prettier` support for semi-colon and once we have that we can migrate `gulp-metal` and all our projects to this configuration.

**Comment by fernandosouza** [comment]
Sure, @eduardolundgren. I'm following the issue and the repo as well.

**Comment by jbalsas** [comment]
Hey @eduardolundgren, I'm not sure I understand the semicolon issue.

Google Style Guide states that:

> Every statement must be terminated with a semicolon. Relying on automatic semicolon insertion is forbidden. [*]

 [\[*\] 4.3.2 Semicolons are required](https://google.github.io/styleguide/jsguide.html#formatting-semicolons-are-required)

Other than that, I'm pretty happy with this setup. Should we also try to enforce the usage of a [Pre-commit hook](https://github.com/prettier/prettier#pre-commit-hook-for-changed-files)?

**Comment by eduardolundgren** [comment]
@jbalsas Yes, prettier doesn't require semicolons, in reality it removes the semicolon, that's why they will add an option for that.

**Comment by jbalsas** [comment]
I actually think the rather enforce it very strictly... https://github.com/prettier/prettier/blob/b91d6384a6db769789ecd332cdf3b26e1e4d59c2/src/printer.js#L958-L959

See this [gist](https://prettier.github.io/prettier/#%7B%22content%22%3A%22const%20a%20%3D%202%5Cn%5Cnlet%20foo%20%3D%20(event)%20%3D%3E%20%7B%5Cn%20%20var%20bar%20%3D%20'bar'%3B%3B%3B%3B%3B%5Cn%20%20%5Cn%20%20console.log(bar)%5Cn%20%20%5Cn%20%20return%20%5Cn%20%202%5Cn%7D%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22doc%22%3Afalse%7D%7D)

I understand the issue to be the opposite... people asking to be able to write without semicolons. Where did you see them being removed?

**Comment by eduardolundgren** [comment]
I would disable more roles:

```js
{
  "extends": ["eslint:recommended", "google"],
  "parserOptions": {
      "ecmaVersion": 2017,
      "sourceType": "module"
  },
  "env": {
    "es6": true
  },
  "rules": {
    "arrow-parens": 0, // disabled because of prettier
    "comma-spacing": 0, // disabled because of prettier
    "no-empty": 0, // disable because we would like to allow empty blocks
    "no-console": 0,
    "no-process-env": 0
  }
}
```

**Comment by bryceosterhaus** [comment]
Any updates with this? Prettier released [v1.1](https://github.com/prettier/prettier/releases/tag/1.1.0) recently and it seems to work pretty well, I added it to both devtools and css-transitions using it through [eslint-plugin-prettier](https://github.com/not-an-aardvark/eslint-plugin-prettier).

**Comment by jbalsas** [comment]
Hey @bryceosterhaus, I sent https://github.com/metal/gulp-metal/pull/47 a while ago. I need to update it now that `1.0` has been released and `useTabs` is available 😉 

We also have a guidelines document we plan to present as soon as possible with this and other related `guidelines`.
